File: packages/nlu/swift/Sources/BridgethingNluKit/NluBundleInference.swift
import BridgethingCompanion
import BridgethingSchema
import CoreML
import Foundation
import Nlu

    public enum BundleError: Error, CustomStringConvertible {
        case catalogMismatch(bundle: [String], catalog: [String])

        public var description: String {
            switch self {
            case let .catalogMismatch(bundle, catalog):
                return "bundle intents \(bundle) do not match the companion catalog \(catalog)"
            }
        }
    }

    private let decoder: NluDecoder
    private let model: any NluModelRunning

    public let rejection: NluRejectionPolicy?

    public init(decoder: NluDecoder, model: any NluModelRunning) throws {
        let info = decoder.info()
        guard info.intentNames == NluIntentCatalog.surfaceNames else {
            throw BundleError.catalogMismatch(bundle: info.intentNames, catalog: NluIntentCatalog.surfaceNames)
        }
        self.decoder = decoder
        self.model = model
        self.rejection = info.rejection.map {
            NluRejectionPolicy(inDomainThreshold: $0.inDomainThreshold, clarifyMargin: $0.clarifyMargin)
        }
    }

    public convenience init(bundleDir: URL, model: any NluModelRunning) throws {
        try self.init(decoder: try NluDecoder.load(bundleDir: bundleDir.path), model: model)
    }

    public static func load(
        bundleDir: URL,
        computeUnits: MLComputeUnits = .cpuOnly,
        deferModel: Bool = false
    ) throws -> NluBundleInference {
        let decoder = try NluDecoder.load(bundleDir: bundleDir.path)
        let heads = decoder.info().closedHeadSizes.count
        let model: any NluModelRunning = deferModel
            ? LazyCoreMlModel(bundleDir: bundleDir, closedHeadCount: heads, computeUnits: computeUnits)
            : try CoreMlNluModel(bundleDir: bundleDir, closedHeadCount: heads, computeUnits: computeUnits)
        return try NluBundleInference(decoder: decoder, model: model)
    }

    public func infer(transcript: String) async throws -> NluInferenceOutput {
        let tokens = try decoder.tokenize(transcript: transcript)
        let out = try model.predict(inputIds: tokens.inputIds, attentionMask: tokens.attentionMask)
        let frame = try decoder.decode(
            transcript: transcript,
            tokens: tokens,
            intentLogits: out.intentLogits,
            bioLogits: out.bioLogits,
            closedLogits: out.closedLogits
        )
        return NluInferenceOutput(
            intentLogits: out.intentLogits.map(Double.init),
            inDomainLogit: -Double(out.oodLogit),
            slots: NluSlotMapping.apply(frame.slots)
        )
    }
}

extension NluBundleInference: NluPrewarmable {
    public func prewarm() async {
        guard let prewarming = model as? any NluModelPrewarming else { return }
        await Task.detached(priority: .userInitiated) { try? prewarming.prewarm() }.value
    }
}

File: packages/nlu/kotlin/nlukit/src/main/kotlin/com/bridgething/nlukit/NluBundleInference.kt
package com.bridgething.nlukit

import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import uniffi.nlu.NluDecoder
import uniffi.nlu.NluDecoderInterface

class NluBundleInference(
    private val decoder: NluDecoderInterface,
    private val model: NluModelRunning,
) : NluInferring, NluPrewarmable, AutoCloseable {
    class CatalogMismatch(bundle: List<String>, catalog: List<String>) :
        Exception("bundle intents $bundle do not match the companion catalog $catalog")

    val rejection: NluRejectionPolicy?

    init {
        val info = decoder.info()
        if (info.intentNames != NluIntentCatalog.surfaceNames) {
            throw CatalogMismatch(info.intentNames, NluIntentCatalog.surfaceNames)
        }
        rejection = info.rejection?.let {
            NluRejectionPolicy(inDomainThreshold = it.inDomainThreshold, clarifyMargin = it.clarifyMargin)
        }
    }

    override suspend fun infer(transcript: String): NluInferenceOutput = withContext(Dispatchers.Default) {
        val tokens = decoder.tokenize(transcript)
        val out = model.predict(tokens.inputIds, tokens.attentionMask)
        val frame = decoder.decode(transcript, tokens, out.intentLogits, out.bioLogits, out.closedLogits)
        NluInferenceOutput(
            intentLogits = out.intentLogits.map(Float::toDouble),
            inDomainLogit = -out.oodLogit.toDouble(),
            slots = NluSlotMapping.apply(frame.slots),
        )
    }

    override fun close() {
        (model as? AutoCloseable)?.close()
        (decoder as? AutoCloseable)?.close()
    }

    override suspend fun prewarm() = withContext(Dispatchers.Default) {
        (model as? NluModelPrewarming)?.prewarm()
        Unit
    }

    companion object {
        fun load(
            bundleDir: File,
            modelFile: File = File(bundleDir, "model.tflite"),
            threads: Int = 2,
            deferModel: Boolean = false,
        ): NluBundleInference {
            val decoder = NluDecoder.load(bundleDir.path)
            val build = { LitertNluModel.load(modelFile, threads) }
            return NluBundleInference(decoder, if (deferModel) LazyNluModel(build) else build())
        }
    }
}
